### DISCUSSION TITLE
Sample Connector for basic user authentication

### DISCUSSION THREAD

**Discussion by maxschueler** [OpenManufacturingPlatform/iotcon-opc-ua-connector-dotnet#25] [Ideas]
A connector that is capable of running a basic user authentication (username/password)

**Comment by maxschueler** [comment]
Current solution approach: 
- user credentials are part of the settings file
- opc ua server endpoints are mapped to user identities (1:1)

The following points are still strongly dependent on the use case in the current solution approach:
- settings file contains unencrypted user credentials
- only one user per endpoint is possible and is statically defined
- no dynamic change in the mapping from endpoint to user identity during runtime possible  (without connector restart)

**Reply by hermotech** [reply]
Another option providing more flexibility was suggested by @ivomp :

A token request can be be sent by the user
This may contain encrypted credentials.

The connector auths with the OPC UA server with the given credentials and return a token to the user.

**Reply by ivomp** [reply]
Besides the problem of specifying a username and password in a settings file. @htbmw also pointed to a potential issue in the way we connect to OPC UA Server endpoints. Which we discovered after having trouble connecting to a test Milo server.

The connector tries to connect to all endpoints of the server until 1 endpoint is successful. But in the case of Milo, Milo refused the connection (even anonymous) due to the fact that the connection attempts was to fast in succession. Adding a delay between attempts fixed the issue.

This behavior makes us question the mechanism. Instead we propose a 'Connect' type command to initiate the connection where the client should specify the endpoint and all relevant credentials needed. Once a connection is established the connector responds with token/key that all subsequent request to the server should be sent along.
